### PR TITLE
Allow custom wait delays

### PR DESCRIPTION
**Pull request by vadimkerr**
# What this PR does

Allows custom wait durations for:
* `Wait` escalation policy
* `>X alerts per Y minutes` escalation policy
* `Wait` user notification policy

## Which issue(s) this PR closes

Related to https://github.com/grafana/oncall/issues/2464

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] Added the relevant release notes label (see labels prefixed w/ `release:`). These labels dictate how your PR will
    show up in the autogenerated release notes.

### CODE REVIEW

**Review by maskin25 — APPROVED** [review]

File: grafana-plugin/src/components/Policy/EscalationPolicy.tsx [hunk 1]
```diff
@@ -27,23 +28,22 @@ import { Schedule } from 'models/schedule/schedule.types';
...
 import { getEscalationPolicyStyles } from './EscalationPolicy.styles';
+import { POLICY_DURATION_LIST_MINUTES, POLICY_DURATION_LIST_SECONDS } from './Policy.consts';
 import { PolicyNote } from './PolicyNote';
 
 interface ElementSortableProps extends WithStoreProps {
   index: number;
 }
 
-export interface EscalationPolicyProps extends ElementSortableProps, Themeable2 {
+interface EscalationPolicyBaseProps {
```
[the remark this answers is not in the document]
Comment (teodosii) [reply grafana-plugin/src/components/Policy/EscalationPolicy.tsx:44]: There's 2 props interfaces basically. One for the consumer side, when you actually render the component, all these are mandatory/optional:

```jsx
data: EscalationPolicyType;
isDisabled?: boolean;
channels?: any[];
onChange: (id: EscalationPolicyType['id'], value: Partial<EscalationPolicyType>) => void;
onDelete: (data: EscalationPolicyType) => void;
escalationChoices: any[];
number: number;
backgroundClassName?: string;
backgroundHexNumber?: string;
isSlackInstalled: boolean;
```

The other one is what we extended due to `ElementSortableProps, Themeable2, WithStoreProps` which don't need to get passed down by the consumers.

If I understood correctly you want these merged? But the result isn't the desired one, unless I'm missing something. I want to expose only what's required to be passed in

File: grafana-plugin/src/components/Policy/Policy.consts.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export const CUSTOM_SILENCE_VALUE = -100;
+
+export const SILENCE_DURATION_LIST: SelectableValue[] = [
+  { value: CUSTOM_SILENCE_VALUE, label: 'Custom' },
+  { value: 1800, label: '30 minutes' },
+  { value: 3600, label: '1 hour' },
+  { value: 7200, label: '2 hours' },
+  { value: 21600, label: '6 hours' },
+  { value: 43200, label: '12 hours' },
+  { value: 86400, label: '24 hours' },
```
[the remark this answers is not in the document]
Comment (teodosii) [reply grafana-plugin/src/components/Policy/Policy.consts.ts:40]: Can change it

File: grafana-plugin/src/pages/incidents/parts/IncidentSilenceModal.tsx [hunk 3]
```diff
@@ -31,66 +39,118 @@ const IncidentSilenceModal: React.FC<IncidentSilenceModalProps> = ({
...
-    setDate(date);
+  function onFormSubmit() {
+    onSave(durationToMilliseconds(parseDuration(getValues('duration'))) / 1000);
+  }
+
+  function onDateChange(newDate: DateTime) {
     const duration = intervalToAbbreviatedDurationString({
       start: new Date(),
-      end: new Date(date.toDate()),
+      end: new Date(newDate.toDate()),
```
[the remark this answers is not in the document]
Comment (teodosii) [reply grafana-plugin/src/pages/incidents/parts/IncidentSilenceModal.tsx:140]: `newDate` is actually a `DateTime` time from grafana.

```jsx
import {
  DateTime,
  ...
} from '@grafana/data';

...

function onDateChange(newDate: DateTime)
```

File: grafana-plugin/src/components/Policy/Policy.consts.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { SelectableValue } from '@grafana/data';
+
+const POLICY_DURATION_LIST: SelectableValue[] = [
```
Comment (vadimkerr) [inline grafana-plugin/src/components/Policy/Policy.consts.ts:3]: I think there also should be a 60 minutes options to match this with the existing backend options

**Review by matiasb — APPROVED** [review]
LGTM

**Comment by mderynck** [comment]
Can we have stricter validation? I was able to enter 2000000000 and -2000000000 resulting in pretty strange escalation snapshots

**Comment by teodosii** [comment]
@mderynck sure, will add it on the frontend. right now it's only checking for valid numbers

**Comment by vadimkerr** [comment]
> Can we have stricter validation? I was able to enter 2000000000 and -2000000000 resulting in pretty strange escalation snapshots

Good catch! I'll add backend validation for this. @teodosii do you have any min and max limits in mind? If not, I'd propose something like minimum is 1 minute and maximum is 24 hours, wdyt?

**Comment by teodosii** [comment]
@vadimkerr I think these limits make sense, anything bigger than that is rather difficult to read in minutes (like 10800 it's a week). 1 and 1440 sound good to me

File: engine/apps/api/views/alert_group.py [hunk 5]
```diff
@@ -643,6 +643,8 @@ def silence(self, request, pk=None):
     )
     @action(methods=["get"], detail=False)
     def silence_options(self, request):
+        # TODO: DEPRECATED, REMOVE IN A FUTURE RELEASE
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/api/views/alert_group.py:646]: yes, that's right

File: grafana-plugin/src/components/Policy/Policy.consts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (teodosii) [reply grafana-plugin/src/components/Policy/Policy.consts.ts]: Done :+1:

**Pull request merged**